src/ProxyProtocol.cc

```diff
@@ -0,0 +1,326 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include <algorithm>
+#include "parser/BinaryTokenizer.h"
+#include "parser/Tokenizer.h"
+#include "ProxyProtocol.h"
+#include "sbuf/StringConvert.h"
+#include "sbuf/Stream.h"
+#include "SquidConfig.h"
+#include "SquidString.h"
+#include "StrList.h"
+
+#if HAVE_SYS_SOCKET_H
+#include <sys/socket.h>
+#endif
+#if HAVE_NETINET_IN_H
+#include <netinet/in.h>
+#endif
+#if HAVE_NETINET_IP_H
+#include <netinet/ip.h>
+#endif
+
+/// magic octet prefix for PROXY protocol version 1
+static const SBuf Proxy1p0magic("PROXY ", 6);
+
+ProxyProtocol::Message::FieldMap ProxyProtocol::Message::PseudoHeaderFields = {
+    { SBuf(":version"), ProxyProtocol::Two::PP2_PSEUDO_VERSION },
+    { SBuf(":command"), ProxyProtocol::Two::PP2_PSEUDO_COMMAND },
+    { SBuf(":src_addr"), ProxyProtocol::Two::PP2_PSEUDO_SRC_ADDR },
+    { SBuf(":dst_addr"), ProxyProtocol::Two::PP2_PSEUDO_DST_ADDR },
+    { SBuf(":src_port"), ProxyProtocol::Two::PP2_PSEUDO_SRC_PORT },
+    { SBuf(":dst_port"), ProxyProtocol::Two::PP2_PSEUDO_DST_PORT }
+};
+
+/// magic octet prefix for PROXY protocol version 2
+static const SBuf Proxy2p0magic("\x0D\x0A\x0D\x0A\x00\x0D\x0A\x51\x55\x49\x54\x0A", 12);
+
+ProxyProtocol::Message::Message(const char *ver, const uint8_t cmd)
```

Comment:
Please use this much more common and lower-maintenance one-initialization-per-line format:

```
name(...):
    member1(...),
    member2(...),
    memberN(...)
```